### PR TITLE
OvS P4runtime client support for linux networking version 2

### PR DESCRIPTION
**Pull request by n-sandeep**
This change includes support for below tables:
 Configure other new P4 tables for
 vxlan_encap (V4 and V6)
 vxlan_encap_pop_vlan (V4 and V6)
 vxlan_decap (V4 and V6)
 vxlan_decap_push_vlan (V4 and V6)
 tunnel_term (V4 and V6)
 rx_tunnel (V4 and V6)
 vlan_push
 vlan_pop
 tunnel_src_port
 vsi_src_port

Actions for these table programming is dependent on the port VLAN mode as well.
Based on VLAN mode, we either remove vlan from packet or add vlan to the packet
Currently we support NATIVE_UNTAGGED and NATIVE_TAGGED vlan modes

### CODE REVIEW

**Comment by n-sandeep** [comment]
> This breaks the DPDK build.
> 
> See the details for the **P4 Control Plane CI Pipeline / build_dpdk_p4cp (pull_request)** check at the bottom of the Conversation page.
> 
> The CI workflow did its job... 🙂

Yes Derek, i am working on that. DPDK build brken due to changes in ES2K. Will raise a new patch by tomorrow

File: ovs-p4rt/ovs_p4rt.cc [hunk 1]
```diff
@@ -113,107 +113,73 @@ void PrepareFdbTxVlanTableEntry(p4::v1::TableEntry* table_entry,
...
+      {
+        auto param = action->add_params();
+        param->set_param_id(GetParamId(p4info, L2_FWD_TX_TABLE_ACTION_L2_FWD,
+                                       ACTION_L2_FWD_PARAM_PORT));
 #if defined(DPDK_TARGET)
-      auto port_id = learn_info.vln_info.vlan_id - 1;
+        auto port_id = learn_info.vln_info.vlan_id - 1;
 #elif defined(ES2K_TARGET)
-      auto port_id = learn_info.mac_addr[1] + ES2K_VPORT_ID_OFFSET;
+        auto port_id = learn_info.mac_addr[1] + ES2K_VPORT_ID_OFFSET;
```
Comment (nupuruttarwar) [inline ovs-p4rt/ovs_p4rt.cc:173]: Here port_id is forwarding vsi_id and we are using a workaround to get the vsi id from 2nd octet of mac address + offest (16). Please add a comment explaining this and also a TODO to revisit this if IMC logic for VSI is changes

File: ovs-p4rt/ovs_p4rt.cc [hunk 2]
```diff
@@ -562,6 +493,7 @@ void PrepareEncapTableEntry(p4::v1::TableEntry* table_entry,
                                      ACTION_VXLAN_ENCAP_PARAM_SRC_ADDR));
       param->set_value(CanonicalizeIp(tunnel_info.local_ip.ip.v4addr.s_addr));
     }
+    // Done
```
Comment (nupuruttarwar) [inline ovs-p4rt/ovs_p4rt.cc:496]: Can you remove debug comments?

File: ovs-p4rt/ovs_p4rt.cc [hunk 3]
```diff
@@ -621,6 +553,7 @@ void PrepareV6EncapTableEntry(p4::v1::TableEntry* table_entry,
                                      ACTION_VXLAN_ENCAP_V6_PARAM_SRC_ADDR));
       param->set_value(CanonicalizeIpv6(tunnel_info.local_ip.ip.v6addr));
     }
+    // Done
```
Comment (nupuruttarwar) [inline ovs-p4rt/ovs_p4rt.cc:556]: Can you remove debug comments?

**Review by nupuruttarwar — APPROVED** [review]
Minor comments. Logic LGTM. Please fix DPDK build issue before merging

File: ovs-p4rt/ovs_p4rt.cc (hunk 1 above)
Comment (n-sandeep) [reply ovs-p4rt/ovs_p4rt.cc]: I can stiill add a comment regarding port ID

**Review by vsureshkumarp — APPROVED** [review]
LGTM..:)

**Review by 5abeel — APPROVED** [review]
LGTM

**Review by ffoulkes — CHANGES_REQUESTED** [review]
I focused primarily on structure and readability, rather than logic.

The DPDK build error needs to be fixed. Everything else is fodder for a subsequent commit.

You can get fast turnaround on a DPDK build check by doing:

```bash
./make-all.sh --target=dpdk --no-build
cmake --build build -j4 --target ovs_sidecar_o
```

Once that succeeds, you can finish the build by doing:

```bash
cmake --build build -j6
```

File: ovs-p4rt/ovs_p4rt.cc [hunk 4]
```diff
@@ -324,129 +325,172 @@ void PrepareFdbTableEntryforV4Tunnel(p4::v1::TableEntry* table_entry,
       param->set_value(ip_address);
     }
   }
+#endif
 
-#elif defined(ES2K_TARGET)
+#if defined(ES2K_TARGET)
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply ovs-p4rt/ovs_p4rt.cc:330]: Sure, will make changes accordingly

File: ovs-p4rt/ovs_p4rt.cc [hunk 5]
```diff
@@ -324,129 +325,172 @@ void PrepareFdbTableEntryforV4Tunnel(p4::v1::TableEntry* table_entry,
...
-          ACTION_SET_TUNNEL_UNDERLAY_V4_OVERLAY_V4_PARAM_DST_ADDR));
-      std::string ip_address =
-          CanonicalizeIp(learn_info.tnl_info.remote_ip.ip.v4addr.s_addr);
-      param->set_value(ip_address);
+    if (learn_info.tnl_info.local_ip.family == AF_INET &&
+        learn_info.tnl_info.remote_ip.family == AF_INET) {
+      if (learn_info.vlan_info.port_vlan_mode == P4_PORT_VLAN_NATIVE_UNTAGGED) {
+        action->set_action_id(GetActionId(
+            p4info, L2_FWD_TX_TABLE_ACTION_POP_VLAN_SET_TUNNEL_UNDERLAY_V4));
+        {
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply ovs-p4rt/ovs_p4rt.cc:340]: Yes, was replicating similar pattern use by existing code. Will consider this when param call is made into a different API.

File: ovs-p4rt/ovs_p4rt.cc [hunk 6]
```diff
@@ -655,102 +673,707 @@ void PrepareV6EncapTableEntry(p4::v1::TableEntry* table_entry,
...
+}
+
+absl::Status ConfigDecapTableEntry(ovs_p4rt::OvsP4rtSession* session,
+                                   const struct tunnel_info& tunnel_info,
+                                   const ::p4::config::v1::P4Info& p4info,
+                                   bool insert_entry) {
+  p4::v1::WriteRequest write_request;
+  ::p4::v1::TableEntry* table_entry;
+
+  if (insert_entry) {
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply ovs-p4rt/ovs_p4rt.cc:1103]: Sure Derek.. will take it in subsequent changes,

File: ovs-p4rt/ovs_p4rt.cc [hunk 7]
```diff
@@ -834,34 +1453,364 @@ void ConfigFdbTableEntry(struct mac_learning_info learn_info,
...
+                                 bool insert_entry) {
+  /* Unimplemented for DPDK target */
+  return;
+}
+
+void ConfigVlanTableEntry(uint16_t vlan_id, bool insert_entry) {
+  /* Unimplemented for DPDK target */
+  return;
+}
+void ConfigTunnelSrcPortTableEntry(struct src_port_info tnl_sp,
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply ovs-p4rt/ovs_p4rt.cc:1802]: This will be fixed as part of OVS submodule update

**Review by ffoulkes — APPROVED** [review]

**Pull request merged**